File: school_member-1/school_member.m
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

#define max(a, b) ((a) > (b)? (a) : (b))
#define min(a, b) ((a) < (b)? (a) : (b))

typedef struct {
    int year;
    int month;
    int day;
} Date;

Date newDate(int yr, int mn, int dy)
{
    Date date;
    static int length[] = { 0, 31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31 };
    int year, month, day;

    month = max(1, mn);
    month = min(month, 12);

    day = max(1, dy);
    day = min(day, length[month]);

    year = max(1, yr);

    date.year = year;
    date.month = month;
    date.day = day;

    return date;
}

Date Now()
{
    time_t theTime = time(NULL);
    struct tm *aTime = localtime(&theTime);
    Date now;

    now.year = aTime->tm_year + 1900;
    now.month =  aTime->tm_mon + 1;
    now.day = aTime->tm_mday;
    return now;
}

char *Date_String(Date date)
{
    char *buf = (char *)malloc(11);
    sprintf(buf, "%.4d-%.2d-%.2d", date.year, date.month, date.day);
    return buf;
}

#import <Foundation/Foundation.h>

@interface Member: NSObject
{
   char *name;
   Date birthdate;
}
- (id) name: (char *) n birthdate: (Date) d;
- (char *) name;
- (int) age;
- (char *) greet;
@end

@implementation Member
- (id) name: (char *) n birthdate: (Date) b
  {
    name = n;
    birthdate = b;
    return self;
  }

- (char *) name
  {
    return name;
  }
  
- (int) age;
  {
    return Now().year - birthdate.year;
  }

- (char *) greet
  {
    char *buf = (char *)malloc(BUFSIZ);
    sprintf(buf, "Name: %s, Age: %d", [self name], [self age]);
    return buf;
  }
@end

@interface Teacher: Member
{
   char *lecture;
}
- (id) name: (char *) n birthdate: (Date) d teaches: (char *) t;
- (char *) teaches;
- (char *) greet;
@end

@implementation Teacher
- (id) name: (char *) n birthdate: (Date) d teaches: (char *) t
  {
    [super name: n birthdate: d];
    lecture = t;
    return self;
  }

- (char *) teaches
  {
    return lecture;
  }

- (char *) greet
  {
    char *buf = (char *)malloc(BUFSIZ);
    char *etc = (char *)malloc(BUFSIZ/4);
    strcpy(buf, [super greet]);
    sprintf(etc, ", Teaches: %s", [self teaches]);
    strcat(buf, etc);
    free(etc);
    return buf;
  }
@end

@interface Student: Member
{
   int id_no;
}
- (id) name: (char *) n birthdate: (Date) d id_no: (int) i;
- (int) id_no;
- (char *) greet;
@end

@implementation Student
- (id) name: (char *) n birthdate: (Date) d id_no: (int) i
  {
    [super name: n birthdate: d];
    id_no = i;
    return self;
  }

- (int) id_no
  {
    return id_no;
  }

- (char *) greet
  {
    char *buf = (char *)malloc(BUFSIZ);
    char *etc = (char *)malloc(BUFSIZ/4);
    strcpy(buf, [super greet]);
    sprintf(etc, ", ID_no: %d", [self id_no]);
    strcat(buf, etc);
    free(etc);
    return buf;
  }
@end

@interface Visiting: Student
{
   Date valid_thru; 
}
- (id) name: (char *) n birthdate: (Date) d id_no: (int) i validThru: (Date) v;
- (bool) expired;
- (char *) greet;
@end

@implementation Visiting
- (id) name: (char *) n birthdate: (Date) d id_no: (int) i validThru: (Date) v
  {
    [super name: n birthdate: d id_no: i];
    valid_thru = v;
    return self;
  }

- (bool) expired
  {
    Date now = Now();
    return now.year >= valid_thru.year;
  }

- (char *) greet
  {
    char *buf = (char *)malloc(BUFSIZ);
    char *etc = (char *)malloc(BUFSIZ/4);
    strcpy(buf, [super greet]);
    sprintf(etc, ", Valid_thru: %s", Date_String(valid_thru));
    strcat(buf, etc);
    free(etc);
    return buf;
  }
@end

int ageComp(const void *e1, const void *e2)
{
    id m1 = *((id*)e1);
    id m2 = *((id*)e2);
    return [m1 age] - [m2 age];
}

int reverseNameComp(const void *e1, const void *e2)
{
    id m1 = *((id*)e1);
    id m2 = *((id*)e2);
    return strcmp((char *)[m2 name], (char *)[m1 name]);
}

int main(int argc, char *argv[])
{
    NSAutoreleasePool *pool = [[NSAutoreleasePool alloc] init];
    id members[10];
    int m_count = 0;
    members[m_count++] = [[[Teacher alloc] init] name: "MH" birthdate: newDate(1971, 12, 7) teaches: "Programming Languages"];
    members[m_count++] = [[[Teacher alloc] init] name: "JY" birthdate: newDate(1975, 9, 21) teaches: "Forbidden Archeology"];
    members[m_count++] = [[[Student alloc] init] name: "YK" birthdate: newDate(1999, 3, 16) id_no: 2051];
    members[m_count++] = [[[Student alloc] init] name: "SH" birthdate: newDate(2000, 10, 5) id_no: 4968];
    members[m_count++] = [[[Visiting alloc] init] name: "Alice" birthdate: newDate(2000, 10, 5) id_no: 9595 validThru: newDate(2019, 12, 25)];
    members[m_count++] = [[[Visiting alloc] init] name: "Vanessa" birthdate: newDate(1999, 3, 27) id_no: 9598 validThru: newDate(2019, 2, 28)];

    int i;
    printf("A few CAU members...\n");
    for (i = 0; i < m_count; i++) {
        char *string_rep = [members[i] greet];
        printf("%s\n", string_rep);
        free(string_rep);
    }
    printf("\n");

    qsort(members, m_count, sizeof(members[0]), ageComp);
    printf("Members sorted on age...\n");
    for (i = 0; i < m_count; i++) {
        char *string_rep = [members[i] greet];
        printf("%s\n", string_rep);
        free(string_rep);
    }
    printf("\n");

    qsort(members, m_count, sizeof(members[0]), reverseNameComp);
    printf("Members reverse sorted on name...\n");
    for (i = 0; i < m_count; i++) {
        char *string_rep = [members[i] greet];
        printf("%s\n", string_rep);
        free(string_rep);
    }
    printf("\n");

    [pool drain];
    return 0;
}

File: school_member-1/school_member.cxx
#define _CRT_SECURE_NO_WARNINGS
#include <iostream>
#include <time.h>
#include <list>
#include <sstream>
#include <iomanip>
#include <iterator>

using namespace std;

// Class Date
inline int max(int a, int b)
{
    if (a > b) return(a); else return (b);
}

inline int min(int a, int b)
{
    if (a > b) return(b); else return (a);
}

class Date
{
public:
    Date();
    Date(int yr, int mn, int dy);
    int Year();
    int Month();
    int Day();
    ~Date();
    string Greet();
    friend Date Now();
private:
    int month, day, year;
    int DaysSoFar();
};

Date::Date()
{
   month = day = year = 1;
}

Date::Date(int yr, int mn, int dy)
{
   static int length[] = { 0, 31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31 };

   month = max(1, mn);
   month = min(month,12);

   day = max(1,dy);
   day = min(day, length[month]);

   year = max(1, yr);
}

Date Now()
{
    time_t theTime = time(NULL);
    struct tm *aTime = localtime(&theTime);
    return Date(aTime->tm_year + 1900, aTime->tm_mon + 1, aTime->tm_mday);
}

Date::~Date()
{
}

int Date::Year()
{
   return year;
}

int Date::Month()
{
   return month;
}

int Date::Day()
{
    return day;
}

string Date::Greet()
{
    stringstream buf;
    buf << to_string(this->Year()) << "-" << setw(2) << setfill('0')
        << to_string(this->Month()) << "-"
        << to_string(this->Day());
    return buf.str();
}
// End of Class Date

class Member
{
public:
    string Name();
    int Age();
    virtual string Greet();
protected:
    Member(string name, Date birthdate);
private:
    string name;
    Date birthdate;
};

Member::Member(string name, Date birthdate)
{
    this->name = name;
    this->birthdate = birthdate;
}

string Member::Name()
{
    return this->name;
}

int Member::Age()
{
    return Now().Year() - this->birthdate.Year();
}

string Member::Greet()
{
    return "Name: " + this->Name() + ", Age: " + std::to_string(this->Age());
}

class Teacher: public Member
{
public:
    Teacher(string name, Date birthdate, string teaches);
    string Teaches();
    string Greet();
private:
    string lecture;
};

Teacher::Teacher(string name, Date birthdate, string teaches): Member(name, birthdate)
{
    this->lecture = teaches;
}

string Teacher::Teaches()
{
    return this->lecture;
}

string Teacher::Greet()
{
    return Member::Greet() + ", Teaches: " + this->Teaches();
}

class Student: public Member
{
public:
    Student(string name, Date birthdate, int id_no);
    int ID_no();
    string Greet();
private:
    int id_no;
};

Student::Student(string name, Date birthdate, int id_no): Member(name, birthdate)
{
    this->id_no = id_no;
}

int Student::ID_no()
{
    return this->id_no;
}

string Student::Greet()
{
    return Member::Greet() + ", ID_no: " + std::to_string(this->ID_no());
}

class Visiting: public Student
{
public:
    Visiting(string name, Date birthdate, int id_no, Date valid_thru);
    bool Expired();
    string Greet();
private:
    Date valid_thru;
    bool expired;
};

Visiting::Visiting(string name, Date birthdate, int id_no, Date valid_thru): Student(name, birthdate, id_no)
{
    this->valid_thru = valid_thru;
}

bool Visiting::Expired() // Buggy...
{
    Date now = Now();
    return now.Year() >= valid_thru.Year();
}

string Visiting::Greet()
{
    return Student::Greet() + ", Valid_thru: " + valid_thru.Greet();
}

// Sorters
bool ageComp(Member *m1, Member *m2)
{
    return m1->Age() < m2->Age();
}

bool reverseNameComp(Member *m1, Member *m2)
{
    return m1->Name().compare(m2->Name()) > 0;
}

int main(int argc, char *argv[])
{
    list<Member*> members;
    members.push_back(new Teacher("MH", Date(1971, 12,  7), "Programming Languages"));
    members.push_back(new Teacher("JY", Date(1975,  9, 21), "Forbidden Archeology"));
    members.push_back(new Student("YK", Date(1999,  3, 16), 2051));
    members.push_back(new Student("SH", Date(2000, 10,  5), 4968));
    members.push_back(new Visiting("Alice", Date(1995, 7, 14), 9595, Date(2019, 12, 25)));
    members.push_back(new Visiting("Vanessa", Date(1998, 3, 27), 9598, Date(2019, 2, 28)));

    cout << "A few CAU members..." << endl;
    list<Member*>::iterator it;
    for (it = members.begin(); it != members.end(); ++it) {
        Member *member = *it;
        cout << member->Greet() << endl;
    }
    cout << endl;

    members.sort(ageComp);
    cout << "Members sorted on age..." << endl;
    for (it = members.begin(); it != members.end(); ++it) {
        Member *member = *it;
        cout << member->Greet() << endl;
    }
    cout << endl;

    members.sort(reverseNameComp);
    cout << "Members reverse sorted on name..." << endl;
    for (it = members.begin(); it != members.end(); ++it) {
        Member *member = *it;
        cout << member->Greet() << endl;
    }

    return 0;
}
